pymc3/distributions/multivariate.py

```diff
@@ -20,133 +21,93 @@
-            self._n = self.chol_cov.shape[-1]
+            if self.chol_cov.ndim != 2:
+                raise ValueError('`chol` must be two dimensional.')
+            deltaop = MvNormalLogp(False)
+            def deltalogp(delta):
+                return deltaop(self.chol_cov, delta)
+            self.deltalogp = deltalogp
 
-    def _quaddist(self, value):
-        """Compute (x - mu).T @ Sigma^-1 @ (x - mu) and the logdet of Sigma."""
-        mu = self.mu
-        if value.ndim > 2 or value.ndim == 0:
-            raise ValueError('Invalid dimension for value: %s' % value.ndim)
-        if value.ndim == 1:
-            onedim = True
-            value = value[None, :]
         else:
-            onedim = False
-
-        delta = value - mu
-
-        if self._cov_type == 'cov':
-            # Use this when Theano#5908 is released.
-            # return MvNormalLogp()(self.cov, delta)
-            dist, logdet, ok = self._quaddist_cov(delta)
-        elif self._cov_type == 'tau':
-            dist, logdet, ok = self._quaddist_tau(delta)
-        else:
-            dist, logdet, ok = self._quaddist_chol(delta)
-
-        if onedim:
-            return dist[0], logdet, ok
-        return dist, logdet, ok
-
-    def _quaddist_chol(self, delta):
-        chol_cov = self.chol_cov
-        _, k = delta.shape
-        k = pm.floatX(k)
-        diag = tt.nlinalg.diag(chol_cov)
-        # Check if the covariance matrix is positive definite.
-        ok = tt.all(diag > 0)
-        # If not, replace the diagonal. We return -inf later, but
-        # need to prevent solve_lower from throwing an exception.
-        chol_cov = tt.switch(ok, chol_cov, 1)
-
-        delta_trans = self.solve_lower(chol_cov, delta.T).T
-        quaddist = (delta_trans ** 2).sum(axis=-1)
-        logdet = tt.sum(tt.log(diag))
-        return quaddist, logdet, ok
+            self._cov_type = 'tau'
+            tau = tt.as_tensor_variable(tau)
+            try:
+                self.chol_tau = tt.slinalg.Cholesky(lower=True, on_error="nan")(tau)
+                def deltalogp(delta):
+                    delta_trans = tt.dot(delta, self.chol_tau)
+                    quaddist = (delta_trans ** 2).sum(axis=-1)
+                    diag = tt.ExtractDiag(view=True)(self.chol_tau)
+                    logdet = -tt.log(diag).sum()
+                    k = delta.shape[-1].astype(theano.config.floatX)
+                    norm = - 0.5 * k * floatX(np.log(2 * np.pi))
+                    ok = ~tt.isnan(self.chol_tau[0,0])
+                    logp = norm - 0.5 * quaddist - logdet
+                    return ifelse(ok, floatX(logp), floatX(-np.inf * tt.ones_like(logp)))
+                self.deltalogp = deltalogp
+            except ValueError:
+                raise_from(ValueError('`tau` must be two dimensional.'), None)
+            self.tau = tau
 
-    def _quaddist_cov(self, delta):
-        return self._quaddist_chol(delta)
+class _QuadFormBase(Continuous, _CovSet):
```

Comment:
We can use `MvNormalLogp` _only_ in `MvNormal.logp_sum`, because it returns the sum of all logp values, not the elementwise logp. (we need the elemwise logp in `pm.compare` etc), so we still need `_quaddist` in MvNormal and StudentT.

Reply:
Ah! That's the bit I was missing! Ok, and do we *want* a GraphOp for the regular elementwise logp?

Reply:
I wrote that some time ago because at least back then it was quite a bit faster. But this might have changed in recent theano versions.

Reply:
Sorry, misread your question. The GraphOp thing is faster for the sum of logp. I don't know for sure if the same thing would work for the elemwise logp, but if I remember correctly, it doesn't. That's why MvNormalLogp isn't used at the moment. We only added `logp_sum` recently and I haven't looked into this since then.